src/state-access.h

```diff
@@ -414,22 +414,41 @@ class state_access : public i_state_access<state_access, pma_entry> {
         return m_m.get_state().htif.iyield;
     }
 
-    NO_INLINE uint64_t do_poll_console(uint64_t mcycle) {
-        const bool htif_console_getchar = static_cast<bool>(read_htif_iconsole() & (1 << HTIF_CONSOLE_GETCHAR));
-        if (htif_console_getchar) {
-            const uint64_t warp_cycle = rtc_time_to_cycle(read_clint_mtimecmp());
-            if (warp_cycle > mcycle) {
-                constexpr uint64_t cycles_per_us = RTC_CLOCK_FREQ / 1000000; // CLOCK_FREQ / 10^6
-                const uint64_t wait = (warp_cycle - mcycle) / cycles_per_us;
-                const int64_t start = os_now_us();
-                os_poll_tty(wait);
-                const int64_t end = os_now_us();
-                const uint64_t elapsed_us = static_cast<uint64_t>(std::max(end - start, INT64_C(0)));
-                const uint64_t tty_cycle = mcycle + (elapsed_us * cycles_per_us);
-                mcycle = std::min(std::max(tty_cycle, mcycle), warp_cycle);
+    NO_INLINE std::pair<uint64_t, bool> do_poll_external_interrupts(uint64_t mcycle, uint64_t mcycle_max) {
+        bool interrupt_raised = false;
+        const bool has_htif_console = m_m.has_htif_console();
+        const bool has_virtio_devices = m_m.has_virtio_devices();
+        const bool has_virtio_console = m_m.has_virtio_console();
+        // Only poll external interrupts if we are in interactive mode (console is enabled or have VirtIO devices)
+        if (unlikely(has_htif_console || has_virtio_devices)) {
+            // Convert the relative interval of cycles we can wait to the interval of host time we can wait
+            uint64_t timeout_us = (mcycle_max - mcycle) / RTC_CYCLES_PER_US;
+            int64_t start_us = 0;
+            if (timeout_us > 0) {
+                start_us = os_now_us();
+            }
+            device_state_access da(*this, mcycle);
+            // Poll virtio for events (e.g console stdin, network sockets)
+            // Timeout may be decremented in case a device has deadline timers (e.g network device)
+            if (has_virtio_devices && has_virtio_console) { // VirtIO + VirtIO console
+                m_m.poll_virtio_devices(&timeout_us, &da);
+                // VirtIO console device will poll TTY
+            } else if (has_virtio_devices) { // VirtIO without a console + HTIF console
```

Comment:
Having a network, but no console is indeed a thing to run scripts. This is something I tested and should work.

The logic changed a bit in the next commits. But it is fine calling `os_poll_tty` even when console is disabled, because it will return immediately since TTY was not initialized.

Reply:
It's just for clarity.

Reply:
Added some new comments here, with more clarifications of all possible scenarios.